README.md

```diff
@@ -0,0 +1,492 @@
+
+---
+
+# Intro
+
+Avenger is a data fetching and caching layer written in TypeScript. Its API is designed to mirror the principles of **Command Query Responsibility Segregation** and facilitate their adoption (if you are new to the concept you can get a grasp of its foundations in [this nice article](https://martinfowler.com/bliki/CQRS.html) by Martin Fowler).
+
+Building user interfaces is a complex task, mainly because of its `IO` intensive nature. Reads (**queries**) and updates (**commands**) toward "external" data sources are ubiquitous and difficult to orchestrate but _orchestration_ is not the only challenge a UI developer faces, _performance_ and _scalability_ are also key aspects of good design.
+
+We believe that an _effective and powerful abstraction to handle caching and synchronization of external data in a declarative way_ is of fundamental importance when designing a solid user interface.
+
+This is what **Avenger** wants to be: an abstraction layer over external data that handles caching and synchronization for you:
+
+!["cached flow"](docs/Avenger.svg)
+
+By separating how we fetch external data and how we update it we are able to state in a very declarative and _natural_ way the correct lifecycle of that data:
+
+```ts
+// define a cached query, with strategy "available" (more about this later)
+const user = queryShallow((id: string) => API.fetchUser(id), available);
+// define a command that invalidates the previous query
+const updateUsername = command((patch: Partial<User>) => API.updateUser(patch), { user })
+
+// declare it for usage in a React component
+const queries = declareQueries({ user });
+const DisplayUsername = queries(props => (
+  <div>
+    {props.queries.fold(
+      () => 'loading...',
+      () => 'error while retrieving user',
+      queries => (
+        <UserNameForm
+          value={queries.user.username}
+          onSubmit={updateUsername}
+        />
+      )
+    )}
+  </div>
+));
+
+// render the component
+<Username queries={{ user: '42' }} />;
+```
+
+# Avenger
+At the very heart of Avenger's DSL there are two constructors: **query** and **command**.
+
+## queries
+The [**`query`**](#query) function allows you to query your data source and get an object of type [**`CachedQuery`**](#CachedQuery) in return.
+It accepts two parameters: the first is a function with a [**`Fetch`**](#Fetch) signature that is used to retrieve data from your data source; the second is an object with the[**`Strategy`**](#Strategy) signature that will be used to decide if the data stored by **Avenger** is still relevant or needs to be refetched.
+
+Although important, `query` is a pretty low-level API and **Avenger** offers some convenient utils with a [**`StrategyBuilder`**](#StrategyBuilder) signature that you should prefer over it (unless you have very specific needs):
+
+- **refetch:** runs the fetch function every time the data is requested.
```

Comment:
I would add: "(unless there's an ongoing pending request, which is always reused)"